### PR TITLE
Implement build versioning for AppVeyor Windows builds

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
Alternative approach to #3649. One thing to note though: all builds in AppVeyor matrix will have the same version number, so installer will not be able to replace, say, SDL1 version by SDL2 version produced during the same build run.

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
Need to check the compatibility of these changes with Visual Studio, in particular regarding `ResourceCompile` phase (rc.exe). It is possible to use `AdditionalOptions` instead of `PreprocessorDefinitions` for this phase if something goes wrong.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , could you please explain why do we need this change from user perspective? We have archives to be used by developers if they want to test different versions. I can see that certain builds might be needed for testing purposes but AppVeyors stores all artifacts for the last 1 month.

**Comment by ihhub** [comment]
Artifacts in AppVeyor could be downloaded using [this explanation](https://github.com/ihhub/fheroes2/wiki/Download-Windows-latest-compiled-packages-of-a-pull-request).

**Comment by oleg-derevenetz** [comment]
Hi @ihhub That's because Inno Setup installer checks file version in Windows PE header, and don't update files which version doesn't change. So currently there no way to install latest version of fheroes2 (from some build artifact) over some previous (currently installed) version just by running installer.

**Comment by ihhub** [comment]
> Hi @ihhub That's because Inno Setup installer checks file version in Windows PE header, and don't update files which version doesn't change. So currently there is no way to install latest version of fheroes2 (from some build artifact) over some previous (currently installed) version just by running installer because all of them have the same version - 0.9.4.0.

I think I got an idea. The issue I see is that we depend on AppVeyor and this build number will be present on release builds.

**Comment by oleg-derevenetz** [comment]
> The issue I see is that we depend on AppVeyor and this build number will be present on release builds.

1. Is this really an issue? Version number will be like this: 0.9.4.8765, so all major version numbers will be still in place.
2. If this is an issue, then is there any way to differentiate AppVeyor release build from regular builds?

**Comment by ihhub** [comment]
> > The issue I see is that we depend on AppVeyor and this build number will be present on release builds.
> 
> 1. Is this really an issue? Version number will be like this: 0.9.4.8765, so all major version numbers will be still in place.
> 2. If this is an issue, then is there any way to differentiate AppVeyor release build from regular builds?

These aren't big issues. I don't think that there's a way to separate release builds from normal builds as right now release is done manually (copy compilned packages and put them into release).

I think we're okay with this hacky approach but all these changes are for development team rather than for users :) Maybe just allowing to override will be even better?

**Comment by ihhub** [comment]
> > Maybe just allowing to override will be even better?
> 
> That's why there is #3649 also :) So it's up to you. It's just that I personally encountered such a problem during the update process and decided to fix it somehow, and got two ideas.

Let me please think a bit over night then :)

**Comment by oleg-derevenetz** [comment]
BTW, there will be also handy for a user to include this info in bug reports - user should just right-click on exe file and we get file version with build number linked to specific commit.

File: appveyor.yml [hunk 1]
```diff
@@ -49,6 +49,13 @@ install:
 init:
   - git config --global core.autocrlf true
 
+# src/fheroes2/system/version.h: BUILD_VERSION will be set to %build_version% inside VisualStudio/common.props
+# src/resources/fheroes2.rc: BUILD_VERSION is limited to 16 bit
+#
+# double percent sign is stripped by AppVeyor, so we can't calculate modulo in the usual way
+before_build:
+  - cmd: set /a build_version = %APPVEYOR_BUILD_NUMBER% - %APPVEYOR_BUILD_NUMBER% / 32768 * 32768
```
Comment (ihhub) [inline appveyor.yml:57]: One thing which bothers me in these changes is this environment variable. Could we please rename it to something less generic like `fheroes2_build_version` not to mess with something else?

**Review by ihhub — COMMENTED** [review]
Hi @oleg-derevenetz , I left just one comment. Could you please take a look?

File: appveyor.yml (hunk 1 above)
Comment (oleg-derevenetz) [reply appveyor.yml:57]: Yes, of course. It really looks too generic. May be ci_build_version?

Comment (ihhub) [reply appveyor.yml:57]: It works too but my prefence with `fheroes2_build_version`. The decision is yours :)

**Comment by ihhub** [comment]
@oleg-derevenetz , thank you very much for these changes!

**Pull request merged**